File: src/main/java+kotlin/com/group4/chatapp/services/AttachmentService.java
package com.group4.chatapp.services;

import com.group4.chatapp.dtos.messages.MessageSendDto;
import com.group4.chatapp.exceptions.ApiException;
import com.group4.chatapp.models.Attachment;
import com.group4.chatapp.repositories.AttachmentRepository;

import org.springframework.stereotype.Service;
import org.springframework.util.CollectionUtils;
import org.springframework.http.HttpStatus;

import java.util.List;
import java.util.Map;
import java.util.Objects;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class AttachmentService {

    private final FileTypeService fileTypeService;
    private final FileStorageService fileStorageService;

    private final AttachmentRepository attachmentRepository;

    public List<Attachment> getAttachments(MessageSendDto dto) {

        var attachments = dto.getAttachments();
        if (CollectionUtils.isEmpty(attachments)) {
            return List.of();
        }

        var uploadedFiles = fileStorageService.uploadMultipleFiles(attachments);

        if (CollectionUtils.isEmpty(uploadedFiles)) {
            return List.of();
        }

        var anySuccess = uploadedFiles.stream()
            .anyMatch(file -> "success".equals(file.get("status")));

        if (!anySuccess) {
            var firstError = uploadedFiles.stream()
                .map(file -> file.get("message"))
                .filter(Objects::nonNull)
                .map(Object::toString)
                .findFirst()
                .orElse("Upload failed");

            throw new ApiException(HttpStatus.BAD_REQUEST, firstError);
        }

        return uploadedFiles
            .stream()
            .map((file) -> {

                var isSuccess = file.get("status").equals("success");
                if (!isSuccess) {
                    return null;
                }

                var resourceType = (String) file.get("resource_type");
                var source = (String) file.get("secure_url");
                var format = (String) file.get("format");

                var type = fileTypeService.checkTypeInFileType(resourceType, format);

                var attachment = Attachment.builder()
                    .source(source)
                    .type(type)
                    .build();

                return attachmentRepository.save(attachment);

            })
            .filter(Objects::nonNull)
            .toList();
    }
}

File: src/main/java+kotlin/com/group4/chatapp/services/AttachmentService.kt
package com.group4.chatapp.services

import com.group4.chatapp.dtos.messages.MessageSendDto
import com.group4.chatapp.exceptions.ApiException
import com.group4.chatapp.models.Attachment
import com.group4.chatapp.repositories.AttachmentRepository
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service

@Service
class AttachmentService(
    private val fileTypeService: FileTypeService,
    private val s3Service: S3Service,
    private val attachmentRepository: AttachmentRepository
) {

    fun getAttachmentOrThrow(id: Long): Attachment =
        attachmentRepository.findById(id).orElseThrow {
            ApiException(HttpStatus.NOT_FOUND, "Attachment not found")
        }

    fun getAttachments(dto: MessageSendDto): List<Attachment> {
        val multipartFiles = dto.attachmentsOrEmpty()
        if (multipartFiles.isEmpty()) {
            return emptyList()
        }

        val uploadedResults = s3Service.uploadMany(multipartFiles)
        if (uploadedResults.isEmpty()) {
            return emptyList()
        }

        val hasSuccess = uploadedResults.any { it is S3Service.UploadResult.Success }
        if (!hasSuccess) {

            val firstError = uploadedResults
                .filterIsInstance<S3Service.UploadResult.Failure>()
                .firstOrNull()
                ?.message
                ?: "Upload failed"

            throw ApiException(HttpStatus.BAD_REQUEST, firstError)
        }

        return uploadedResults
            .asSequence()
            .filterIsInstance<S3Service.UploadResult.Success>()
            .map { result ->
                val attachmentType = fileTypeService.checkTypeInFileType(
                    result.resourceType,
                    result.format
                )

                Attachment.of(result.secureUrl, attachmentType)
            }
            .map(attachmentRepository::save)
            .toList()
    }
}
